hs/src/Reach/Optimize.hs

```diff
@@ -246,6 +248,99 @@ opt_v2v v = fst <$> opt_v2p v
 opt_v2a :: DLVar -> App DLArg
 opt_v2a v = snd <$> opt_v2p v
 
+class IsCom a where
+  isCom :: a -> Maybe (DLStmt, a)
+
+instance IsCom DLTail where
+  isCom = \case
+    DT_Com m k -> Just (m, k)
+    _ -> Nothing
+
+instance IsCom CTail where
+  isCom = \case
+    CT_Com m k -> Just (m, k)
+    _ -> Nothing
+
+instance IsCom ETail where
+  isCom = \case
+    ET_Com m k -> Just (m, k)
+    _ -> Nothing
+
+instance IsCom LLStep where
+  isCom = \case
+    LLS_Com m k -> Just (m, k)
+    _ -> Nothing
+
+instance IsCom LLConsensus where
+  isCom = \case
+    LLC_Com m k -> Just (m, k)
+    _ -> Nothing
+
+class LiftPure a where
+  -- Returns what `DLVar` is `DL_Set` to and the pure `DLStmt`s along the way
+  -- Returns `Nothing` if there are impure stmts in `a` or there is no `DL_Set`
+  liftPure :: DLVar -> a -> (Maybe DLArg, [DLStmt])
+
+instance LiftPure DLTail where
+  liftPure dv = \case
+    DT_Return _ -> (Nothing, [])
+    DT_Com (DL_Set _ lhs rhs) (DT_Return _)
+      | lhs == dv -> (Just rhs, [])
+    DT_Com ds dt ->
+      case isPure ds of
+        True -> do
+          let (ma, dt') = liftPure dv dt
+          (ma, ds : dt')
+        False -> (Nothing, [])
+
+floatVar :: (DLStmt -> t -> t) -> SrcLoc -> DLVar -> t -> App t
+floatVar mkk at dv k = do
+  eVarsSet <- (liftIO . readIORef) =<< asks eVarsSet
+  case M.lookup dv eVarsSet of
+    -- There were zero branches that could not be lifted, i.e.
+    -- we generated a `const` assign, now drop this `var`.
```

Comment:
It is possible for this to happen because even `Optimize` is generally "top down", `float` is "locally bottom up" and we float after optimizing the tail